src/backend/dx12ll/src/data.rs

```diff
@@ -129,40 +129,36 @@ pub fn map_format(format: Format, is_target: bool) -> Option<DXGI_FORMAT> {
     })
 }
 
-pub fn map_resource_state(resource_state: ResourceState) -> D3D12_RESOURCE_STATES {
-    // TODO: A huge puzzle \o/
+//TODO: map_buffer_resource_stat {
+    //handle D3D12_RESOURCE_STATE_INDEX_BUFFER
+//}
+
+pub fn map_image_resource_state(access: ImageAccess, layout: ImageLayout) -> D3D12_RESOURCE_STATES {
     let mut state = D3D12_RESOURCE_STATE_COMMON;
 
-    if resource_state.contains(memory::INDEX_BUFFER_READ) {
-        state = state | D3D12_RESOURCE_STATE_INDEX_BUFFER;
-    }
-    if resource_state.contains(memory::VERTEX_BUFFER_READ) ||
-       resource_state.contains(memory::CONSTANT_BUFFER_READ)  {
-        state = state | D3D12_RESOURCE_STATE_VERTEX_AND_CONSTANT_BUFFER;
+    if access.contains(memory::RENDER_TARGET_CLEAR) {
+        state = state | D3D12_RESOURCE_STATE_RENDER_TARGET;
     }
-    if resource_state.contains(memory::INDIRECT_COMMAND_READ) {
-        state = state | D3D12_RESOURCE_STATE_INDIRECT_ARGUMENT;
+
+    if access.contains(memory::RESOLVE_SRC) {
+        state = state | D3D12_RESOURCE_STATE_RESOLVE_SOURCE;
     }
-    if resource_state.contains(memory::RENDER_TARGET_CLEAR) {
-        state = state | D3D12_RESOURCE_STATE_RENDER_TARGET;
+    if access.contains(memory::RESOLVE_DST) {
+        state = state | D3D12_RESOURCE_STATE_RESOLVE_DEST;
     }
-    /*
-    if resource_state.contains(memory::TRANSFER_SRC) {
+
+    if access.contains(memory::TRANSFER_READ) {
         state = state | D3D12_RESOURCE_STATE_COPY_SOURCE;
     }
-    if resource_state.contains(memory::TRANSFER_DST) {
+    if access.contains(memory::TRANSFER_WRITE) {
         state = state | D3D12_RESOURCE_STATE_COPY_DEST;
     }
-    */
-    if resource_state.contains(memory::RESOLVE_SRC) {
-        state = state | D3D12_RESOURCE_STATE_RESOLVE_SOURCE;
-    }
-    if resource_state.contains(memory::RESOLVE_DST) {
-        state = state | D3D12_RESOURCE_STATE_RESOLVE_DEST;
-    }
 
-    if resource_state.contains(memory::PRESENT) {
-        state = state | D3D12_RESOURCE_STATE_PRESENT;
+    match layout {
+        ImageLayout::Present => {
+            state = state | D3D12_RESOURCE_STATE_PRESENT;
```

Comment:
`D3D12_RESOURCE_STATE_PRESENT` is _not_ a flag, is actually 0!
https://msdn.microsoft.com/de-de/library/windows/desktop/dn986744(v=vs.85).aspx